File: core/src/main/java/fr/sncf/osrd/infra_state/implementation/standalone/StandaloneDetectionSectionState.java
package fr.sncf.osrd.infra_state.implementation.standalone;

import fr.sncf.osrd.infra.api.reservation.DetectionSection;
import fr.sncf.osrd.infra.api.reservation.ReservationRoute;
import fr.sncf.osrd.infra_state.api.DetectionSectionState;
import fr.sncf.osrd.infra_state.api.ReservationTrain;

public class StandaloneDetectionSectionState implements DetectionSectionState {

    private final Summary summary;
    private final ReservationRoute route;
    private final ReservationTrain occupyingTrain;
    private final DetectionSection detectionSection;

    /** Constructor */
    public StandaloneDetectionSectionState(
            Summary summary,
            ReservationRoute route,
            ReservationTrain occupyingTrain,
            DetectionSection detectionSection
    ) {
        this.summary = summary;
        this.route = route;
        this.occupyingTrain = occupyingTrain;
        this.detectionSection = detectionSection;
        assert summary.equals(Summary.FREE) == (occupyingTrain == null)
                : "occupying train must be null if and only if section is free";
        assert summary.equals(Summary.FREE) == (route == null)
                : "route must be null if and only if section is free";
    }

    @Override
    public Summary summarize() {
        return summary;
    }

    @Override
    public ReservationRoute getRoute() {
        return route;
    }

    @Override
    public ReservationTrain getOccupyingTrain() {
        return occupyingTrain;
    }

    @Override
    public DetectionSection getDetectionSection() {
        return detectionSection;
    }
}

File: core/src/main/kotlin/fr/sncf/osrd/dyn_infra/implementation/standalone/StandaloneDetectionSectionState.kt
package fr.sncf.osrd.dyn_infra.implementation.standalone

import fr.sncf.osrd.infra.api.reservation.ReservationRoute
import fr.sncf.osrd.dyn_infra.api.ReservationTrain
import fr.sncf.osrd.infra.api.reservation.DetectionSection
import fr.sncf.osrd.dyn_infra.api.DetectionSectionState

class StandaloneDetectionSectionState(
    private val summary: DetectionSectionState.Summary,
    override val route: ReservationRoute?,
    override val occupyingTrain: ReservationTrain?,
    override val detectionSection: DetectionSection?
) : DetectionSectionState {

    /** Constructor  */
    init {
        assert(summary == DetectionSectionState.Summary.FREE == (occupyingTrain == null)) { "occupying train must be null if and only if section is free" }
        assert(summary == DetectionSectionState.Summary.FREE == (route == null)) { "route must be null if and only if section is free" }
    }

    override fun summarize(): DetectionSectionState.Summary {
        return summary
    }
}